pkg/rlptools/wasm_utils.go

```diff
@@ -10,119 +10,193 @@ import (
 )
 
 var (
 	WASMFilterImageURL = common.FetchEnv("RELATED_IMAGE_WASMSHIM", "oci://quay.io/kuadrant/wasm-shim:latest")
 )
 
-type GatewayAction struct {
-	Configurations []kuadrantv1beta1.Configuration `json:"configurations"`
+// WasmRules computes WASM rules from the policy and the targeted Route (which can be nil when a gateway is targeted)
+func WasmRules(rlp *kuadrantv1beta2.RateLimitPolicy, route *gatewayapiv1alpha2.HTTPRoute) []wasm.Rule {
+	rules := make([]wasm.Rule, 0)
+	if rlp == nil {
+		return rules
+	}
 
-	// +optional
-	Rules []kuadrantv1beta1.Rule `json:"rules,omitempty"`
+	for limitName, limit := range rlp.Spec.Limits {
+		// 1 RLP limit <---> 1 WASM rule
+		limitFullName := FullLimitName(rlp, limitName)
+		rule := ruleFromLimit(limitFullName, &limit, route)
+
+		rules = append(rules, rule)
+	}
+
+	return rules
 }
 
-func DefaultGatewayConfiguration(key client.ObjectKey) []kuadrantv1beta1.Configuration {
-	return []kuadrantv1beta1.Configuration{
-		{
-			Actions: []kuadrantv1beta1.ActionSpecifier{
-				{
-					GenericKey: &kuadrantv1beta1.GenericKeySpec{
-						DescriptorValue: key.String(),
-						// using default value as specified in Envoy spec
-						// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#config-route-v3-ratelimit-action-generickey
-						DescriptorKey: &[]string{"ratelimitpolicy"}[0],
-					},
-				},
-			},
-		},
+func ruleFromLimit(limitFullName string, limit *kuadrantv1beta2.Limit, route *gatewayapiv1alpha2.HTTPRoute) wasm.Rule {
+	if limit == nil {
+		return wasm.Rule{}
+	}
+
+	return wasm.Rule{
+		Conditions: conditionsFromLimit(limit, route),
+		Data:       dataFromLimt(limitFullName, limit),
 	}
 }
 
-// GatewayActionsFromRateLimitPolicy return flatten list from GatewayAction from the RLP
-func GatewayActionsFromRateLimitPolicy(rlp *kuadrantv1beta1.RateLimitPolicy, route *gatewayapiv1alpha2.HTTPRoute) []GatewayAction {
-	flattenActions := make([]GatewayAction, 0)
-	if rlp == nil {
-		return flattenActions
+func conditionsFromLimit(limit *kuadrantv1beta2.Limit, route *gatewayapiv1alpha2.HTTPRoute) []wasm.Condition {
+	if limit == nil {
+		return make([]wasm.Condition, 0)
 	}
 
-	for idx := range rlp.Spec.RateLimits {
-		// Skip those RateLimit objects with empty configurations, even if they have rules defined
-		if len(rlp.Spec.RateLimits[idx].Configurations) == 0 {
-			continue
+	// TODO(eastizle): review this implementation. This is a first naive implementation.
+	// The conditions must always be a subset of the route's matching rules.
+
+	conditions := make([]wasm.Condition, 0)
+
+	for routeSelectorIdx := range limit.RouteSelectors {
+		// TODO(eastizle): what if there are only Hostnames (i.e. empty "matches" list)
+		for matchIdx := range limit.RouteSelectors[routeSelectorIdx].Matches {
+			condition := wasm.Condition{
+				AllOf: patternExpresionsFromMatch(&limit.RouteSelectors[routeSelectorIdx].Matches[matchIdx]),
+			}
```

Comment:
Actually, the Wasm conditions should not come directly from the limit, but use the limit's `routeSelectors` to select the rules from within the HTTPRoutes: https://github.com/Kuadrant/kuadrant-operator/issues/200